[LoopVectorize] Use new getUniqueLatchExitBlock routine

david-arm opened the pull request:
With PR #88385 I am introducing support for vectorising more loops with early exits that don't require a scalar epilogue. As such, if a loop doesn't have a unique exit block it will not automatically imply we require a scalar epilogue. Also, in all places in the code today where we use the variable LoopExitBlock we actually mean the exit block from the latch. Therefore, it seemed reasonable to add a new getUniqueLatchExitBlock that allows the caller to determine the exit block taken from the latch and use this instead of getUniqueExitBlock. I also renamed LoopExitBlock to be LatchExitBlock. I feel this not only better reflects how the variable is used today, but also prepares the code for PR #88385.

While doing this I also noticed that one of the comments in requiresScalarEpilogue is wrong when we require a scalar epilogue, i.e. when we're not exiting from the latch block. This doesn't always imply we have multiple exits, e.g. see the test in

Transforms/LoopVectorize/unroll_nonlatch.ll

where the latch unconditionally branches back to the only exiting block.

Dinistro reviewed (commented):
Looks reasonable, but I'm not too familiar with LoopVectorize, so someone else should accept this.

david-arm committed:
[LoopVectorize] Use new getUniqueLatchExitBlock routine

With PR #88385 I am introducing support for vectorising more
loops with early exits that don't require a scalar epilogue.
As such, if a loop doesn't have a unique exit block it will
not automatically imply we require a scalar epilogue. Also,
in the only place in the code today where we use the variable
LoopExitBlock we actually mean the exit block from the latch.
Therefore, it seemed reasonable to add a new
getUniqueLatchExitBlock helper that allows the caller to
determine the exit block taken from the latch and use this
instead of getUniqueExitBlock. I also removed LoopExitBlock
since it was only used in one place.

While doing this I also noticed that one of the comments in
requiresScalarEpilogue is wrong when we require a scalar
epilogue, i.e. when we're not exiting from the latch block.
This doesn't always imply we have multiple exits, e.g. see
the test in

Transforms/LoopVectorize/unroll_nonlatch.ll

where the latch unconditionally branches back to the only
exiting block.

david-arm commented:
Rebase

david-arm commented:
Ping

david-arm committed:
Address review comment

* Use find_singleton in getUniqueLatchExitBlock.

david-arm commented on llvm/include/llvm/Support/GenericLoopInfoImpl.h:
```diff
@@ -159,6 +159,16 @@ BlockT *LoopBase<BlockT, LoopT>::getUniqueExitBlock() const {
+BlockT *LoopBase<BlockT, LoopT>::getUniqueLatchExitBlock() const {
+  const BlockT *Latch = getLoopLatch();
+  assert(Latch && "Latch block must exists");
+  SmallVector<BlockT *, 4> ExitBlocks;
```
Done!

huntergr-arm reviewed (approved):
LGTM

fhahn commented on llvm/include/llvm/Support/GenericLoopInfo.h:
```diff
@@ -294,6 +294,10 @@ template <class BlockT, class LoopT> class LoopBase {
   BlockT *getUniqueExitBlock() const;
 
+  /// Return the unique exit block for the latch, or null if there are multiple
+  /// different exit blocks.
```
or the latch is not exiting?

fhahn commented on llvm/include/llvm/Support/GenericLoopInfoImpl.h:
```diff
@@ -159,6 +159,17 @@ BlockT *LoopBase<BlockT, LoopT>::getUniqueExitBlock() const {
+BlockT *LoopBase<BlockT, LoopT>::getUniqueLatchExitBlock() const {
+  BlockT *Latch = getLoopLatch();
+  assert(Latch && "Latch block must exists");
+  auto IsExitBlock = [&](BlockT *BB, bool AllowRepeats) -> BlockT * {
```
nit: sufficient to capture `this`?
```suggestion
  auto IsExitBlock = [this](BlockT *BB, bool AllowRepeats) -> BlockT * {
```

fhahn reviewed (approved):
LGTM, looks like this helps to nicely avoid code changes  in follow-up changes.

david-arm committed:
Address review comments

david-arm commented:
Looks like the build is failing due to an issue with git checkout. I ran "make check-all" downstream and it seems to pass.
